landoapi/hgexports.py

```diff
@@ -71,12 +75,92 @@ def _no_line_breaks(break_string: str) -> str:
     return "".join(break_string.strip().splitlines())
 
 
-class HgPatchHelper(object):
-    """Helper class for parsing Mercurial patches/exports."""
+def parse_git_author_information(user_header: bytes) -> tuple[bytes, bytes]:
+    """Parse user's name and email address from a Git style author header.
+
+    Converts a header like 'User Name <user@example.com>' to its separate parts.
+    """
+    name, email = parseaddr(user_header.decode("utf-8"))
+    return name.encode("utf-8"), email.encode("utf-8")
+
+
+def get_timestamp_from_git_date_header(date_header: bytes) -> str:
+    """Convert a Git patch date header into a timestamp."""
+    header_datetime = dateparse(date_header)
+    return str(math.floor(header_datetime.timestamp()))
+
+
+def get_timestamp_from_hg_date_header(date_header: bytes) -> bytes:
+    """Return the first part of the `hg export` date header.
+
+    >>> get_timestamp_from_hg_date_header(b"1686621879 14400")
+    b"1686621879"
+    """
+    return date_header.split(b" ")[0]
+
+
+class PatchHelper(object):
+    """Base class for parsing patches/exports."""
 
     def __init__(self, fileobj: io.BytesIO):
         self.patch = fileobj
         self.headers = {}
+
+    @staticmethod
+    def _is_diff_line(line: bytes) -> bool:
+        return DIFF_LINE_RE.search(line) is not None
+
+    def header(self, name: bytes | str) -> Optional[bytes]:
+        """Returns value of the specified header, or None if missing."""
+        name = name.encode("utf-8") if isinstance(name, str) else name
+        return self.headers.get(name.lower())
+
+    def set_header(self, name: bytes | str, value: bytes):
+        """Set the header `name` to `value`."""
+        name = name.encode("utf-8") if isinstance(name, str) else name
+        self.headers[name.lower()] = value
+
+    def commit_description(self) -> bytes:
+        """Returns the commit description."""
+        raise NotImplementedError("`commit_description` not implemented.")
+
+    def get_diff(self) -> bytes:
+        """Return the patch diff."""
+        raise NotImplementedError("`get_diff` not implemented.")
+
+    def write_commit_description(self, f: io.BytesIO):
+        """Writes the commit description to the specified file object."""
+        f.write(self.commit_description())
+
+    def write_diff(self, file_obj: io.BytesIO):
+        """Writes the diff to the specified file object."""
+        file_obj.write(self.get_diff())
+
+    def write(self, f: io.BytesIO):
+        """Writes whole patch to the specified file object."""
+        try:
+            while 1:
+                buf = self.patch.read(16 * 1024)
+                if not buf:
+                    break
+                f.write(buf)
+        finally:
+            self.patch.seek(0)
```

Comment:
I've made a [previous comment](https://github.com/mozilla-conduit/lando-api/pull/328#discussion_r1270973325) about a similar implementation/piece of code, there are some improvements there that could be applied here.

Reply:
I'll handle this change in [bug 1846330](https://bugzilla.mozilla.org/show_bug.cgi?id=1846330), it's out of scope for this PR. This code is pre-existing in Lando and is just being moved to the new base `PatchHelper`.